(PUP-8409) Enable config cmd to print conf file compatible format

kris-bosland opened the pull request:
When showing a setting with config print, without this change values are returned formatted for internal use.
With this change, returned values are formatted like the conf file expects.

haus commented on lib/puppet/settings.rb:
```diff
@@ -1294,7 +1294,7 @@ def lookup(name)
     # @return [Object] The configuration setting value or nil if the setting is not known
     # @api public
-    def interpolate(name)
+    def interpolate(name, to_string=false)
```
As you've added a new param here, it should get added to the yarddoc above the method.

puppetcla commented:
CLA signed by all contributors.

joshcooper commented on lib/puppet/settings.rb:
```diff
@@ -1321,6 +1326,16 @@ def interpolate(name)
+        value.join(', ')
+      else
+        setting.munge(value)
+      end
```
What about pushing this logic into the respective settings classes?

joshcooper commented on lib/puppet/settings.rb:
```diff
@@ -1312,7 +1313,11 @@ def interpolate(name)
+            interpolate_as_string(val, setting)
+          else
+            setting.munge(val)
+          end
```
What about moving most of the `interpolate` logic to a private method (say `_interpolate` and then have two public methods `interpolate` and `print` (or `render` or ...?) that call the private method, eg

```ruby
def interpolate(name)
  setting, val = _interpolate(name)
  setting.munge(val)
end

def print(name)
  setting, val = _interpolate(name)
  setting.print(val)
end

def _interpolate(name)
  # does most of the work
  ...
  setting, val
end
```

joshcooper commented:
conflicts need to be resolved

kris-bosland commented:
Updated: @joshcooper, @jhelwig , @haus , and @ericldelaney if you are still looking.
I have implemented @joshcooper suggestions, but skipped the array test because I wasn't sure where to put that functionality in the new scheme.  In any case, for that further level of detail we should file/fork a new ticket and look at a broader refactoring (for example, taking a look at to_config and how it relates to what we want from config print).

haus commented on lib/puppet/settings.rb:
```diff
@@ -1292,35 +1292,46 @@ def lookup(name)
-    def interpolate(name)
+    def interpolate(name, to_string=false)
       setting = @defaults[name]
+      return nil if not setting
```
I think `return nil unless setting` is a more common way of writing this.

haus commented on lib/puppet/settings.rb:
```diff
@@ -1292,35 +1292,46 @@ def lookup(name)
-      end
+    def print(name)
+      setting = @defaults[name]
+      return nil if not setting
```
Same as above wrt `unless` instead of `if not`

joshcooper commented on lib/puppet/settings.rb:
```diff
@@ -1292,35 +1292,46 @@ def lookup(name)
     # @return [Object] The configuration setting value or nil if the setting is not known
     # @api public
-    def interpolate(name)
+    def interpolate(name, to_string=false)
```
I think we can drop `to_string` here.

puppetcla commented:
CLA signed by all contributors.

joshcooper commented on spec/unit/face/config_spec.rb:
```diff
@@ -23,6 +23,19 @@ module PuppetFaceSpecs
+  end
+
+  it "prints arrays correctly" do
+    skip "Still doesn't print arrays like they would appear in config"
```
I think we want to use `pending` here so that the test will fail if the code is later changed to print arrays correctly (so we know which tests to go update)

joshcooper commented:
This isn't working for me, I'm still getting `Infinity`?

```
$ grep timeout ~/.puppetlabs/etc/puppet/puppet.conf
environment_timeout = unlimited
$ bx puppet config print --all --section main | grep environment_timeout
Resolving settings from section 'main' in environment 'production'
environment_timeout = Infinity
$ bx puppet config print environment_timeout --section main
Resolving settings from section 'main' in environment 'production'
Infinity
```

joshcooper commented:
Sorry I just read the commit message now. Could you update it to mention that the environment_timeout defaults were incorrect and how that caused the problem we were seeing above?

joshcooper commented on lib/puppet/settings/environment_conf.rb:
```diff
@@ -38,7 +38,7 @@ def self.load_from(path_to_env, global_module_path)
   # without interpolation.  This is a special case for the default configured
   # environment returned by the Puppet::Environments::StaticPrivate loader.
-  def self.static_for(environment, environment_timeout = 0, static_catalogs = false, rich_data = false)
+  def self.static_for(environment, environment_timeout = 'unlimited', static_catalogs = false, rich_data = false, call_flag=false)
```
I think we want to keep this as `0`, as that is the correct default.

kris-bosland commented:
jenkins please test this on windows2012r2-64

joshcooper commented on lib/puppet/settings/environment_conf.rb:
```diff
@@ -38,7 +38,7 @@ def self.load_from(path_to_env, global_module_path)
   # without interpolation.  This is a special case for the default configured
   # environment returned by the Puppet::Environments::StaticPrivate loader.
-  def self.static_for(environment, environment_timeout = 0, static_catalogs = false, rich_data = false)
+  def self.static_for(environment, environment_timeout = 0, static_catalogs = false, rich_data = false, call_flag=false)
```
It looks like we can drop the `call_flag` parameter and the call site above?

kris-bosland commented:
Good catch of some debugging code, @joshcooper .  Removed.

kris-bosland commented:
Merging myself after checking with @joshcooper.

joshcooper commented on lib/puppet/environments.rb:
```diff
@@ -115,7 +115,7 @@ def get(name)
       env = get(name)
       if env
-        Puppet::Settings::EnvironmentConf.static_for(env, 0, Puppet[:static_catalogs], Puppet[:rich_data])
+        Puppet::Settings::EnvironmentConf.static_for(env, Puppet[:environment_timeout], Puppet[:static_catalogs], Puppet[:rich_data])
```
@hlindberg @thallgren FYI, it looks like this was a bug that hardcoded the environment timeout for static environments to "never cache" instead of whatever the global default was (ideally "unlimited" in production environments). Let us know if I've mischaracterized that.
